// @prisma/photon/dist/runtime/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var dmmf_types_1 = require("./dmmf-types");
exports.DMMF = dmmf_types_1.DMMF;
var dmmf_1 = require("./dmmf");
exports.DMMFClass = dmmf_1.DMMFClass;
var deep_set_1 = require("./utils/deep-set");
exports.deepGet = deep_set_1.deepGet;
exports.deepSet = deep_set_1.deepSet;
var query_1 = require("./query");
exports.makeDocument = query_1.makeDocument;
exports.transformDocument = query_1.transformDocument;
var NodeEngine_1 = require("@prisma/engine-core/dist/NodeEngine");
exports.Engine = NodeEngine_1.NodeEngine;
var debug_1 = require("debug");
exports.debugLib = debug_1.default;
var printDatasources_1 = require("./utils/printDatasources");
exports.printDatasources = printDatasources_1.printDatasources;
var chalk_1 = require("chalk");
exports.chalk = chalk_1.default;
var printStack_1 = require("./utils/printStack");
exports.printStack = printStack_1.printStack;
var mergeBy_1 = require("./mergeBy");
exports.mergeBy = mergeBy_1.mergeBy;
