// @prisma/fetch-engine/dist/src/getos.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const os_1 = __importDefault(require("os"));
const child_process_1 = require("child_process");
const debug_1 = __importDefault(require("debug"));
const debug = debug_1.default('getos');
async function getos() {
    const platform = os_1.default.platform();
    if (platform !== 'linux') {
        return {
            platform,
        };
    }
    return {
        platform: 'linux',
        libssl: await getLibSslVersion(),
    };
}
exports.getos = getos;
async function getLibSslVersion() {
    const [version, ls] = await Promise.all([
        gracefulExec(`openssl version -v`),
        gracefulExec(`ls -l /lib64 | grep ssl;
    ls -l /usr/lib64 | grep ssl`),
    ]);
    debug({ version });
    debug({ ls });
    if (version) {
        const match = /^OpenSSL\s(\d+\.\d+\.\d+)/.exec(version);
        if (match) {
            return match[1];
        }
    }
    if (ls) {
        const match = /libssl\.so\.(\d+\.\d+\.\d+)/.exec(ls);
        if (match) {
            return match[1];
        }
    }
    return undefined;
}
exports.getLibSslVersion = getLibSslVersion;
async function gracefulExec(cmd) {
    return new Promise(resolve => {
        try {
            child_process_1.exec(cmd, (err, stdout, stderr) => {
                resolve(String(stdout));
            });
        }
        catch (e) {
            resolve(undefined);
            return undefined;
        }
    });
}
async function getPlatform() {
    const { platform, libssl } = await getos();
    if (platform === 'darwin') {
        return 'darwin';
    }
    debug({ platform, libssl });
    if (platform === 'linux' && libssl) {
        if (libssl === '1.0.2') {
            return 'linux-glibc-libssl1.0.2';
        }
        if (libssl === '1.0.1') {
            return 'linux-glibc-libssl1.0.1';
        }
    }
    return 'linux-glibc-libssl1.1.0';
}
exports.getPlatform = getPlatform;
