// @prisma/engine-core/dist/log.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
const logLevelMap = {
    CRIT: 'critical',
    ERRO: 'error',
    WARN: 'warning',
    INFO: 'info',
    DEBG: 'debug',
    TRCE: 'trace',
};
function rustToPublicLogLevel(rustLevel) {
    return logLevelMap[rustLevel];
}
function convertLog(rustLog) {
    const { msg, level, application, ts } = rustLog, rest = __rest(rustLog, ["msg", "level", "application", "ts"]);
    return Object.assign({ message: msg, level: rustToPublicLogLevel(level), application: application, date: new Date(ts) }, rest);
}
exports.convertLog = convertLog;
