// nexus-prisma/dist/naming-strategies.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const pluralize_1 = __importDefault(require("pluralize"));
const utils_1 = require("./utils");
exports.defaultArgsNamingStrategy = {
    whereInput(typeName, fieldName) {
        return `${utils_1.upperFirst(typeName)}${utils_1.upperFirst(fieldName)}WhereInput`;
    },
    orderByInput(typeName, fieldName) {
        return `${utils_1.upperFirst(typeName)}${utils_1.upperFirst(fieldName)}OrderByInput`;
    },
    relationFilterInput(typeName, fieldName) {
        return `${utils_1.upperFirst(typeName)}${utils_1.upperFirst(fieldName)}Filter`;
    },
};
exports.defaultFieldNamingStrategy = {
    findOne: (_, modelName) => modelName.toLowerCase(),
    findMany: (_, modelName) => pluralize_1.default(modelName).toLowerCase(),
    create: fieldName => fieldName,
    update: fieldName => fieldName,
    delete: fieldName => fieldName,
    deleteMany: fieldName => fieldName,
    updateMany: fieldName => fieldName,
    upsert: fieldName => fieldName,
};
